File: python/lexical_analyzer.py
import tokens


DIGIT = "digit"
ALPHA = "alpha"


class UnexpectedCharacter(Exception):
  pass


class NoTokenMatched(Exception):
  pass


class State:
  @staticmethod
  def get_character_type(character):
    if character.isdigit():
      return DIGIT
    elif character.isalpha():
      return ALPHA
    return None

  def __init__(self, matched_token_type=None):
    self.transitions = {}
    self.matched_token_type = matched_token_type

  def add_transition(self, character, state):
    self.transitions.update({character: state})

  def next_state(self, character):
    if character in self.transitions:
      return self.transitions[character]
    elif (char_type := State.get_character_type(character)) in self.transitions:
      return self.transitions[char_type]
    elif self.matched_token_type:
      return self.matched_token_type

    if character:
      raise UnexpectedCharacter(character)
    raise NoTokenMatched()
         

def initialize_state_machine():
  states = []
  states.append(State())
  states.append(State(tokens.Integer))
  states.append(State())
  states.append(State(tokens.OpenParen))
  states.append(State(tokens.CloseParen))
  states.append(State(tokens.Minus))
  states.append(State(tokens.Plus))
  states.append(State(tokens.Multiply))
  states.append(State(tokens.Divide))
  states.append(State(tokens.Float))
  states.append(State(tokens.EOL))
  states.append(State(tokens.WhiteSpace))

  states[0].add_transition(DIGIT, states[1])
  states[0].add_transition('.', states[2])
  states[0].add_transition('(', states[3])
  states[0].add_transition(')', states[4])
  states[0].add_transition('-', states[5])
  states[0].add_transition('+', states[6])
  states[0].add_transition('*', states[7])
  states[0].add_transition('/', states[8])
  states[0].add_transition('\n', states[10])
  states[0].add_transition(' ', states[11])
  states[0].add_transition('\t', states[11])

  states[1].add_transition(DIGIT, states[1])
  states[1].add_transition('.', states[2])

  states[2].add_transition(DIGIT, states[9])
  
  states[9].add_transition(DIGIT, states[9])

  states[11].add_transition(' ', states[11])
  states[11].add_transition('\t', states[11])

  return states
  

def parse_tokens(_input):
  text = _input + "\n"
  states = initialize_state_machine()
  current_state = states[0]
  token_start = 0
  char_pos = 0
  token_list = []

  while char_pos < len(text):
    value = None
    try:
      value = current_state.next_state(text[char_pos])
    except UnexpectedCharacter as e:
      error_position = " " * char_pos
      error_position += "^"
      mono_char_text = text.replace('\t', ' ')
      raise UnexpectedCharacter(f"Unexpected character {e.args[0]} at position {char_pos}\n{mono_char_text[:-1]}\n{error_position}")
    except NoTokenMatched as e:
      raise e

    if value is tokens.EOL:
      token_list.append(value())
      break

    if type(value) is type(tokens.Token) and issubclass(value, tokens.Token):
      current_state = states[0]
      token_text = text[token_start:char_pos]
      new_token = None
      if value is tokens.Integer or value is tokens.Float:
        new_token = value(token_text)
      else:
        new_token = value()
      token_list.append(new_token)
      token_start = char_pos
    elif isinstance(value, State):
      current_state = value
      char_pos += 1 
  
  return list(filter(lambda x: not isinstance(x, tokens.WhiteSpace), token_list))
    

File: rust/src/lexical_analyzer.rs
#[cfg(test)]
mod tests;

use crate::tokens::Token;
use std::{error::Error, fmt::Display};

#[derive(Debug, PartialEq)]
pub struct NotAsciiError;

impl Display for NotAsciiError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Expression was not an ASCII string.")
    }
}

impl Error for NotAsciiError {}

#[derive(Debug, PartialEq)]
pub struct InvalidTokenError {
    pub position: usize,
    pub value: String,
}

impl InvalidTokenError {
    pub fn arrow(&self) -> String {
        let mut s = " ".to_owned().repeat(self.position - 1);
        s.push('^');
        s
    }
}

impl Display for InvalidTokenError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Unexpected character {} at position {}",
            self.value, self.position
        )
    }
}

impl Error for InvalidTokenError {}

#[derive(Debug)]
enum State {
    Initial,
    Integer,
    FloatStart,
    OpenParen,
    CloseParen,
    Minus,
    Plus,
    Multiply,
    Divide,
    Float,
    EOL,
    Whitespace,
    Variable,
}

enum CharType {
    Letter,
    Digit,
    Whitespace,
    Other(char),
}

#[derive(Debug)]
pub struct TokenParser {
    current_state: State,
    expression: String,
    position: usize,
}

impl TokenParser {
    pub fn new(expression: &String) -> Result<TokenParser, NotAsciiError> {
        if !expression.is_ascii() {
            Err(NotAsciiError)
        } else {
            Ok(TokenParser {
                current_state: State::Initial,
                expression: expression.clone(),
                position: 0,
            })
        }
    }

    pub fn get_tokens(&mut self) -> Result<Vec<Token>, InvalidTokenError> {
        let mut token_vec: Vec<Token> = vec![];
        while let Some(next) = self.next_token() {
            #[cfg(test)]
            println!("resolved token: {:?}", next);
            if next != Token::Whitespace {
                if let Token::InvalidToken(s) = next {
                    Err(InvalidTokenError {
                        position: self.position + 1,
                        value: s,
                    })?
                } else {
                    token_vec.push(next.clone());
                }
            }
        }
        Ok(token_vec)
    }

    fn next_token(&mut self) -> Option<Token> {
        self.current_state = State::Initial;
        let start = self.position;
        let mut resolved_token = None;
        let mut exp = self.expression.clone();
        exp.push('\n');

        for next_char in exp[start..].chars() {
            let next_state = Self::transition(&self.current_state, &next_char);
            match next_state {
                None => {
                    #[cfg(test)]
                    println!("transition failed, resolving");
                    let potential_token =
                        Self::token_from_state(&self.current_state, &exp[start..self.position]);
                    resolved_token = if let Token::InvalidToken(_) = potential_token {
                        Some(Token::InvalidToken(next_char.to_string()))
                    } else {
                        Some(potential_token)
                    };
                    break;
                }
                Some(state) => {
                    #[cfg(test)]
                    println!("{:?} ({:?})-> {:?}", self.current_state, next_char, state);
                    self.current_state = state;
                }
            }
            self.position += 1;
        }
        resolved_token
    }

    fn token_from_state(state: &State, value: &str) -> Token {
        #[cfg(test)]
        println!("state: {:?}, value: {}", state, value);
        match state {
            State::Initial => Token::InvalidToken(value.to_owned()),
            State::FloatStart => Token::InvalidToken(value.to_owned()),
            State::EOL => Token::EOL,
            State::Variable => Token::new_variable(value),
            State::Integer => Token::new_number(value),
            State::Float => Token::new_number(value),
            State::OpenParen => Token::new_paren(value),
            State::CloseParen => Token::new_paren(value),
            State::Minus => Token::new_op(value),
            State::Plus => Token::new_op(value),
            State::Multiply => Token::new_op(value),
            State::Divide => Token::new_op(value),
            State::Whitespace => Token::Whitespace,
        }
    }

    fn transition(current_state: &State, next_char: &char) -> Option<State> {
        let char_type = Self::get_char_type(*next_char);
        match (current_state, char_type) {
            (State::Initial, CharType::Digit) => Some(State::Integer),
            (State::Initial, CharType::Other('.')) => Some(State::FloatStart),
            (State::Initial, CharType::Other('(')) => Some(State::OpenParen),
            (State::Initial, CharType::Other(')')) => Some(State::CloseParen),
            (State::Initial, CharType::Other('-')) => Some(State::Minus),
            (State::Initial, CharType::Other('+')) => Some(State::Plus),
            (State::Initial, CharType::Other('*')) => Some(State::Multiply),
            (State::Initial, CharType::Other('/')) => Some(State::Divide),
            (State::Initial, CharType::Other('\n')) => Some(State::EOL),
            (State::Initial, CharType::Whitespace) => Some(State::Whitespace),
            (State::Initial, CharType::Letter) => Some(State::Variable),
            (State::Initial, CharType::Other('_')) => Some(State::Variable),
            //
            (State::Integer, CharType::Digit) => Some(State::Integer),
            (State::Integer, CharType::Other('.')) => Some(State::FloatStart),
            //
            (State::FloatStart, CharType::Digit) => Some(State::Float),
            //
            (State::Float, CharType::Digit) => Some(State::Float),
            //
            (State::Whitespace, CharType::Whitespace) => Some(State::Whitespace),
            //
            (State::Variable, CharType::Letter) => Some(State::Variable),
            (State::Variable, CharType::Digit) => Some(State::Variable),
            (State::Variable, CharType::Other('_')) => Some(State::Variable),
            _ => None,
        }
    }

    fn get_char_type(c: char) -> CharType {
        if c.is_ascii_alphabetic() {
            CharType::Letter
        } else if c.is_ascii_digit() {
            CharType::Digit
        } else if Self::is_whitespace(c) {
            CharType::Whitespace
        } else {
            CharType::Other(c)
        }
    }

    fn is_whitespace(c: char) -> bool {
        c == ' ' || c == '\t'
    }
}
